cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java

```diff
@@ -464,12 +465,18 @@ public void apply(MetadataDataset input) throws Exception {
   }
 
   @Override
-  public Set<MetadataRecord> getSnapshotBeforeTime(MetadataScope scope, final Set<Id.NamespacedId> entityIds,
-                                                    final long timeMillis) {
+  public Set<MetadataRecord> getSnapshotBeforeTime(final MetadataScope scope, final Set<Id.NamespacedId> entityIds,
+                                                   final long timeMillis) {
     return execute(new TransactionExecutor.Function<MetadataDataset, Set<MetadataRecord>>() {
       @Override
       public Set<MetadataRecord> apply(MetadataDataset input) throws Exception {
-        return input.getSnapshotBeforeTime(entityIds, timeMillis);
+        ImmutableSet.Builder<MetadataRecord> builder = ImmutableSet.builder();
```

Comment:
better to do only what is required inside the transaction. Let's move the conversion to ``Set<MetadataRecord>`` outside the transaction.